File: leetcode/scala/1.two-sum.scala
// O(n^2) time 
object Solution {
  def twoSum(nums: Array[Int], target: Int): Array[Int] = {
    for (i <- 0 until nums.length) {
      for (j <- (i + 1) until nums.length) {
        if (nums(i) + nums(j) == target) {
          return Array(i, j)
        }
      }
    }

    Array()
  }
}

// O(n) time && space
object Solution_v2 {
  def twoSum(nums: Array[Int], target: Int): Array[Int] = {
    val map = scala.collection.mutable.HashMap[Int, Int]()

    for (i <- 0 until nums.length) {
      if (map.contains(target - nums(i))) {
        return Array(map(target - nums(i)), i)
      }
        map.put(nums(i),i)
    }
    return Array(0, 0)

  }
}

File: leetcode/cpp/1.two-sum.cpp
#include"oj_header.h"

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target)
    {
        unordered_map<int, int> map;

        for (int i = 0; i < nums.size(); ++i) {
            if (map.find(target - nums[i]) != map.end()) {
                return { i, map[target - nums[i]] };
            }
            map[nums[i]] = i;
        }
        return {};
    }
};
